katdal/chunkstore_s3.py

```diff
@@ -398,35 +437,164 @@ class S3ChunkStore(ChunkStore):
+                        raise S3ObjectNotFound(_error_msg(chunk_name, response))
+                    elif self._retries.is_retry(method, status):
+                        raise S3ServerGlitch(_error_msg(chunk_name, response), status)
+                    elif 400 <= status < 600:
+                        raise StoreUnavailable(_error_msg(chunk_name, response))
+                try:
+                    yield response
+                except TruncatedRead as trunc_error:
+                    # A truncated read is considered a glitch with custom status
+                    prefix = 'Chunk {!r}: '.format(chunk_name) if chunk_name else ''
+                    glitch_error = S3ServerGlitch(prefix + str(trunc_error),
+                                                  _TRUNCATED_HTTP_STATUS_CODE)
+                    raise_from(glitch_error, trunc_error)
+
+    def complete_request(self, method, url, chunk_name='',
+                         process=lambda response: None, **kwargs):
+        """Send HTTP request to S3 server, process response and retry if needed.
+
+        This retries temporary HTTP errors, including reset connections while
+        processing a successful response.
+
+        Parameters
+        ----------
+        method, url : str
+            The standard required parameters of :meth:`requests.Session.request`
+        chunk_name : str, optional
+            Name of chunk, used for error reporting only
+        process : function, signature ``result = process(response)``, optional
+            Function that will process response (the default does nothing)
+        kwargs : optional
+            Passed on to :meth:`request` and :meth:`requests.Session.request`
+
+        Returns
+        -------
+        result : object
+            The output of the `process` function applied to a successful response
+
+        Raises
+        ------
+        AuthorisationFailed
+            If the request is not authorised by appropriate token or credentials
+        S3ObjectNotFound
+            If S3 object request fails because it does not exist
+        S3ServerGlitch
+            If S3 object request fails because server is temporarily overloaded
+        StoreUnavailable
+            If a general HTTP error occurred that is not ignored
+        """
+        retries = self._retries.new()
+        while True:
+            try:
+                with self.request(method, url, chunk_name, **kwargs) as response:
+                    result = process(response)
+            except S3ServerGlitch as e:
+                # Retry based on status of response until we run out of retries
+                response = HTTPResponse(status=e.status_code)
+                try:
+                    retries = retries.increment(method, url, response)
+                except MaxRetryError:
+                    # Raise the final straw that broke the retry camel's back
+                    raise e
```

Comment:
What does the stack trace look like from here (in Python 3). Possibly `raise_from(e, None)` would give a more helpful stack trace (i.e. hiding the stack trace of the retry machinery), but my mental model isn't good enough to be sure without trying it.

Reply:
I thought I was hiding it by not using `raise from` but let's see.

Reply:
Can you show an example of what the stack trace looks like?

Reply:
I like your suggestion better, although the current one isn't too bad on truncation errors.

**Current 50x error**

```
MaxRetryError                             Traceback (most recent call last)
~/git/katdal/katdal/chunkstore_s3.py in complete_request(self, method, url, chunk_name, process, **kwargs)
#######
MaxRetryError: None: Max retries exceeded with url: http://127.0.0.1:51481/katdal-unittest/please-respond-with-502-for-1.2-seconds/x/00003.npy (Caused by ResponseError('too many 502 error responses'))

During handling of the above exception, another exception occurred:

S3ServerGlitch                            Traceback (most recent call last)
#######
~/git/katdal/katdal/chunkstore_s3.py in request(self, method, url, chunk_name, ignored_errors, timeout, **kwargs)
    529                         raise S3ObjectNotFound(msg)
    530                     elif self._retries.is_retry(method, status):
--> 531                         raise S3ServerGlitch(msg, status)
    532                     else:
    533                         raise StoreUnavailable(msg)

S3ServerGlitch: "Chunk 'katdal-unittest/please-respond-with-502-for-1.2-seconds/x/00003': Store responded with HTTP error 502 (Suggested by unit test) to request: GET http://127.0.0.1:51481/katdal-unittest/please-respond-with-502-for-1.2-seconds/x/00003.npy"
```

**Current truncation error**

```
TruncatedRead                             Traceback (most recent call last)
~/git/katdal/katdal/chunkstore_s3.py in request(self, method, url, chunk_name, ignored_errors, timeout, **kwargs)
#######
TruncatedRead: Error reading from S3 HTTP response: expected 8 more byte(s), got EOF

The above exception was the direct cause of the following exception:

S3ServerGlitch                            Traceback (most recent call last)
#######
/usr/local/lib/python3.7/site-packages/future/utils/__init__.py in raise_from(exc, cause)
    398         myglobals['__python_future_raise_from_cause'] = cause
    399         execstr = "raise __python_future_raise_from_exc from __python_future_raise_from_cause"
--> 400         exec(execstr, myglobals, mylocals)
    401 
    402     def raise_(tp, value=None, tb=None):

~/git/katdal/katdal/chunkstore_s3.py in <module>

S3ServerGlitch: "Chunk 'katdal-unittest/please-truncate-chunks-for-0.8-seconds/x/00003': Error reading from S3 HTTP response: expected 8 more byte(s), got EOF"
```

The `raise_from(e, None)` gets rid of the top part of both. The bottom part of the traceback is pretty messy in general, with only the final exception message making sense. At least the top part of the truncation error reveals that the numpy header was chopped off.